finetune/base_models/bert/featurizer.py

```diff
@@ -8,6 +8,80 @@
 from finetune.nn.target_blocks import smooth_pos_attn
 from finetune.errors import FinetuneError
 from finetune.nn.auxiliary import dense_with_custom_init
+from finetune.util.metrics import read_eval_metrics
+
+
+def get_improvement_decay(total_num_steps, eval_dir, eval_freq, train):
+    def py_has_improved():
+        print("ENTERING HERE {} {}".format(eval_dir, train))
+        if not train:
+            return False
+        eval_results = {
+            k: v for k, v in
+            read_eval_metrics(eval_dir).items()
+            if "loss" in v # Filter out anything in eval that does not have a loss value                                                               
+        }
+        if len(eval_results) == 0:
+            return False
+        most_recent_eval = max(eval_results.items(), key=lambda x: x[0])  # last steps.                                                               
+        best_eval = min(eval_results.items(), key=lambda x: x[1]["loss"])  # lowest_loss                                                              
+        if most_recent_eval == best_eval:
+            return True
+        return False
+    has_improved = tf.cast(
+        tf.cond(
+            tf.equal((tf.train.get_global_step() - 1) % eval_freq, 0),
+            true_fn=lambda: tf.py_function(func=py_has_improved, inp=[], Tout=tf.bool),
+            false_fn=lambda: False
+        ),
+        tf.float32
+    )
+    decay = tf.get_variable("decay", shape=[], dtype=tf.float32, trainable=False, initializer=tf.zeros_initializer())
+    decay_op = tf.assign(decay, get_decay_for_half(total_num_steps) * has_improved + decay * (1 - has_improved))
```

Comment:
might be preferable to increment one step at a time?  rather than a sharp jump once model gets lower loss at previous %?